pkg/container-disk/container-disk.go

```diff
@@ -357,37 +235,178 @@ func generateContainerFromVolume(vmi *v1.VirtualMachineInstance, config *virtcon
 		Resources: resources,
 		SecurityContext: &kubev1.SecurityContext{
 			RunAsUser:                &userId,
-			RunAsNonRoot:             &nonRoot,
-			AllowPrivilegeEscalation: &noPrivilegeEscalation,
+			RunAsNonRoot:             virtpointer.P(true),
+			AllowPrivilegeEscalation: virtpointer.P(false),
 			Capabilities: &kubev1.Capabilities{
 				Drop: []kubev1.Capability{"ALL"},
 			},
 		},
 	}
+	switch {
+	case isInit:
+		container.Name = toContainerName(volume.Name) + "-init"
+		container.Args = []string{"--no-op"}
+	case isKernelBoot:
+		container.Name = toContainerName(KernelBootVolumeName)
+		container.VolumeMounts = append(container.VolumeMounts, kubev1.VolumeMount{
+			Name:      KernelBootVolumeName,
+			MountPath: PidFileDir,
+		})
+	default:
+		container.Name = toContainerName(volume.Name)
+		container.VolumeMounts = append(container.VolumeMounts, kubev1.VolumeMount{
+			Name:      GetPidfileVolumeName(volume.Name),
+			MountPath: PidFileDir,
+		})
+	}
 
 	return container
 }
 
+func GetPidfileVolumeName(name string) string {
+	return fmt.Sprintf("pidfile-%s", name)
+}
+
+func GetPidfileDir(name string) string {
+	return filepath.Join(mountBaseDir, name)
+}
+
+func GetPidfilePath(name string) string {
+	return filepath.Join(GetPidfileDir(name), Pidfile)
+}
+
+func GetVolumeMountPidfileContainerDisk(name string) k8sv1.VolumeMount {
+	return k8sv1.VolumeMount{
+		Name:      GetPidfileVolumeName(name),
+		MountPath: GetPidfileDir(name),
+	}
+}
+
+func readPidfile(volName string) (int, error) {
+	t, err := os.ReadFile(GetPidfilePath(volName))
+	if err != nil {
+		return -1, err
+	}
+	pid, err := strconv.Atoi(string(t))
+	if err != nil {
+		return -1, err
+	}
+	return pid, nil
+}
+
+func GetContainerDiksPath(volume *v1.Volume) (string, error) {
+	if volume.VolumeSource.ContainerDisk == nil {
+		return "", fmt.Errorf("not a container disk")
+	}
+	pid, err := readPidfile(volume.Name)
+	if err != nil {
+		return "", err
+	}
+	path := volume.VolumeSource.ContainerDisk.Path
+	if path == "" {
+		path = DiskSourceFallbackPath
+	}
+	fullPath := filepath.Join("/proc", strconv.Itoa(pid), "/root", path)
+	files, err := os.ReadDir(fullPath)
+	if err != nil {
+		return "", err
+	}
+	if len(files) == 0 {
+		return "", fmt.Errorf("no file found in folder %s, no disk present", path)
+	} else if len(files) > 1 {
+		return "", fmt.Errorf("more than one file found in folder %s, only one disk is allowed", path)
+	}
+
+	return filepath.Join(fullPath, files[0].Name()), nil
```

Comment:
What if there are multiple files in a containerdisk? Is that a supportable use case? IIRC in CDI you can also choose which disk to import?